aquadoggo/src/materializer/tasks/dependency.rs

```diff
@@ -1,9 +1,293 @@
+        }
+    }
+
+    Ok(Some(next_tasks.into_iter().flatten().collect()))
+}
+
+#[cfg(test)]
+mod tests {
+    use p2panda_rs::document::{DocumentId, DocumentViewId};
+    use p2panda_rs::identity::KeyPair;
+    use p2panda_rs::operation::{
+        AsVerifiedOperation, OperationValue, PinnedRelation, PinnedRelationList, Relation,
+        RelationList,
+    };
+    use p2panda_rs::storage_provider::traits::OperationStore;
+    use p2panda_rs::test_utils::constants::TEST_SCHEMA_ID;
+    use p2panda_rs::test_utils::fixtures::{
+        create_operation, random_document_id, random_document_view_id,
+    };
+    use rstest::rstest;
+
+    use crate::config::Configuration;
+    use crate::context::Context;
+    use crate::db::stores::test_utils::{insert_entry_operation_and_view, test_db, TestSqlStore};
+    use crate::db::traits::DocumentStore;
+    use crate::materializer::tasks::reduce_task;
+    use crate::materializer::TaskInput;
+
+    use super::dependency_task;
+
+    #[rstest]
+    #[case(test_db(1, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("profile_picture", OperationValue::Relation(Relation::new(random_document_id())))],
+        vec![]), 0)]
+    #[case(test_db(1, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("favorite_book_images", OperationValue::RelationList(RelationList::new([0; 6].iter().map(|_|random_document_id()).collect())))],
+        vec![]), 0)]
+    #[case(test_db(1, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("something_from_the_past", OperationValue::PinnedRelation(PinnedRelation::new(random_document_view_id())))],
+        vec![]), 1)]
+    #[case(test_db(1, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("many_previous_drafts", OperationValue::PinnedRelationList(PinnedRelationList::new([0; 2].iter().map(|_|random_document_view_id()).collect())))],
+        vec![]), 2)]
+    #[case(test_db(1, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("one_relation_field", OperationValue::PinnedRelationList(PinnedRelationList::new([0; 2].iter().map(|_|random_document_view_id()).collect()))),
+             ("another_relation_field", OperationValue::RelationList(RelationList::new([0; 6].iter().map(|_|random_document_id()).collect())))],
+        vec![]), 2)]
+    // This document has been updated
+    #[case(test_db(4, 1, false, TEST_SCHEMA_ID.parse().unwrap(),
+        vec![("one_relation_field", OperationValue::PinnedRelationList(PinnedRelationList::new([0; 2].iter().map(|_|random_document_view_id()).collect()))),
+             ("another_relation_field", OperationValue::RelationList(RelationList::new([0; 6].iter().map(|_|random_document_id()).collect())))],
+        vec![("one_relation_field", OperationValue::PinnedRelationList(PinnedRelationList::new([0; 3].iter().map(|_|random_document_view_id()).collect()))),
+             ("another_relation_field", OperationValue::RelationList(RelationList::new([0; 10].iter().map(|_|random_document_id()).collect())))],
+    ), 3)]
```

Comment:
It would be quite helpful to have some comment on each case that documents what the tested assumptions are. The `case()` syntax is very compact and terse, which can make it hard to pick apart what's going on here.

Something like

> plain relation fields should not result in dependency tasks being planned